### ISSUE TITLE
403 access denied at `node weather-app.js` step

### ISSUE THREAD

**Issue by shpros** [Azure-Samples/key-vault-node-getting-started#3]
## TL;DR
I am pretty sure I have followed all of the instructions, **including using the object ID of the service principal** where appropriate, but I get a 403 "access denied" response from `node weather-app.js vaultname city`.

Not sure whether to post this here or on SO.

## Details

I've been trying to get this tutorial to work for several days, but I can't get past "access denied" from the `node weather-app.js vaultname city` step. 

Although I have been working most of the time on Windows, **I have repeated this entire tutorial on Mac OS X El Capitan, with the same 403 "access denied" results.**

What I've done:
1. OS == Windows 7 using MinGW64. Node/npm versions:

  ```
$ node -v
v7.2.1
$ npm -v
3.10.10
  ```

2. Cloned the key-vault-node-getting-started repo.
3. Installed the dependencies.
4. Created an Azure service principal. I have tried this using both the [Azure portal](https://azure.microsoft.com/documentation/articles/resource-group-create-service-principal-portal/) and [PowerShell](https://azure.microsoft.com/documentation/articles/resource-group-authenticate-service-principal/). I call this one "SampleClient" below. This is the one that will get the initial access, with which I will log in to give the other one access.
5. Created another Azure service principal using Azure Portal. I call this one "SampleApp". This is the one to which I will grant access programmatically via index.js.
6. Set environment variables (`export` using MinGW bash shell; `set` using Windows CMD):

```
export AZURE_SUBSCRIPTION_ID=GUID, my subscription ID, d3a4...
export CLIENT_ID=GUID, SampleClient's application ID, d6a9...
export APPLICATION_SECRET=value of SampleClient's secret key or password
export DOMAIN=GUID, directory ID for my organization's Azure AD, 094c...
export OBJECT_ID=GUID, SampleClient's service principal object ID, d62e...

export SP_KEYVAULT_OPERATIONS=GUID, SampleApp's application ID, 3206...
export OBJECT_ID_KEYVAULT_OPERATIONS=GUID, SampleApp's service principal object ID, fdad...
# also tried setting this to SampleApp's application object ID, 459a..., but that gives a 400 "unauthorized_client"
export WEATHER_APP_KEY=value of SampleApp's secret key

export KEYVAULT_SECRET_NAME="open-weather-map-key"
```
7. `node index.js` # success
8. In Windows CMD, because MinGW bash gives `process.stdin.setRawMode is not a function`:

```
azure login -u "d6a9..." --service-principal --tenant "094c..."
password: (give it the value of APPLICATION_SECRET)
# success

azure keyvault secret set --vault-name VAULT_NAME_FROM_index.js --secret-name "open-weather-map-key" SECRET_CREATED_AT_openweathermap.org
# success
```
9. Back in MinGW bash: `node weather-app.js VAULT_NAME_FROM_index.js london`

Results:

```
\PATH\TO\key-vault-node-getting-started\weather-app.js:52
            if (err) throw err;
                     ^

Error: Access denied
    at \PATH\TO\key-vault-node-getting-started\node_modules\azure-keyvault\lib\keyVaultClient.js:3446:19
    at retryCallback (\PATH\TO\key-vault-node-getting-started\node_modules\ms-rest\lib\filters\systemErrorRetryPolicyFilter.js:85:9)
    at retryCallback (\PATH\TO\key-vault-node-getting-started\node_modules\ms-rest\lib\filters\exponentialRetryPolicyFilter.js:135:9)
    at handleRedirect (\PATH\TO\key-vault-node-getting-started\node_modules\ms-rest\lib\filters\redirectFilter.js:36:9)
    at Request._callback (\PATH\TO\key-vault-node-getting-started\node_modules\ms-rest\lib\requestPipeline.js:126:16)
    at Request.self.callback (\PATH\TO\key-vault-node-getting-started\node_modules\ms-rest\node_modules\request\request.js:187:22)
    at emitTwo (events.js:106:13)
    at Request.emit (events.js:191:7)
    at Request.<anonymous> (\PATH\TO\key-vault-node-getting-started\node_modules\ms-rest\node_modules\request\request.js:1044:10)
    at emitOne (events.js:96:13)
```

I have read [this SO question](http://stackoverflow.com/q/40025598/18356) and its answers, particularly the one that points at [this other SO answer](http://stackoverflow.com/a/40002553/18356). I have also looked at [this azure-sdk-for-node issue](https://github.com/Azure/azure-sdk-for-node/issues/1932), which points to the applicationId-vs-servicePrincipalObjectId problem, but I'm already using the sp object ID.

I have run `get-azurermkeyvault -vaultname testkvXXXX` with results as follows:

```
Vault Name                       : testkvXXXX
Resource Group Name              : testrgYYYY
Location                         : westus
Resource ID                      : /subscriptions/d3a4.../resourceGroups/testrgYYYY/providers/Microsoft.KeyVault/vaults/testkvXXXX
Vault URI                        : https://testkvXXXX.vault.azure.net/
Tenant ID                        : 094c...
SKU                              : Standard
Enabled For Deployment?          : False
Enabled For Template Deployment? : False
Enabled For Disk Encryption?     : False
Access Policies                  : 
                                   Tenant ID                   : 094c...
                                   Object ID                   : d62e...
                                   Application ID              : 
                                   Display Name                : SampleClient (with its IdentifierUri)
                                   Permissions to Keys         : Backup, Create, Delete, Get, Import, List, Restore, Update
                                   Permissions to Secrets      : All
                                   Permissions to Certificates : 

                                   Tenant ID                   : 094c...
                                   Object ID                   : fdad...
                                   Application ID              : 3206...
                                   Display Name                : SampleApp (with its IdentifierUri)
                                   Permissions to Keys         : Get, Import, List
                                   Permissions to Secrets      : All
                                   Permissions to Certificates : 
```

Also `Get-AzureRmADServicePrincipal -ObjectId fdad...`:

```
DisplayName                    Type                           ObjectId
-----------                    ----                           --------
SampleClient                   ServicePrincipal               fdad...
```

## Summary

I have tried:
* Two different OSes
* Two different methods of creating the service principal client
* Setting OBJECT_ID_KEYVAULT_OPERATIONS to the application's object ID instead of the service principal's object ID

What am I doing wrong?

**Comment by pomortaz** [comment]
@shpros I think you are not correctly setting your access policy. You are setting application ID, which is used when a the client making request **on behalf of** a principal. It is a special case and I don't think it is what you are trying to do. So only set the **Object ID** and leave **Application ID** empty.

**Comment by shpros** [comment]
@pomortaz Thanks for the reply.

I think you're talking about SampleApp, which is the app to which I've programmatically given access, and which is the one that is trying to get the secret out of the key vault. On my most recent attempt, I used PowerShell to create this app and its service principal, following the [Create Service Principal with Password](https://docs.microsoft.com/en-us/azure/azure-resource-manager/resource-group-authenticate-service-principal#create-service-principal-with-password) instructions.

To reiterate, I'm not running any of my own code, I'm just setting environment variables and following along with this tutorial.

If I'm not mistaken, the access policy for SampleApp is created when I run `node index.js`, using the settings in these lines:

```
        var newAccessPolicyEntry = {
            tenantId: domain,
            objectId: objectIdForKeyVault,
            applicationId: keyVaultSp,
            permissions: {
                keys: ['get', 'list', 'import'],
                secrets: ['all']
            }
        };
```

`objectIdForKeyVault` and `keyVaultSp` are set in index.js as follows:

```
// service principal details that we want give access to the key vault.
var objectIdForKeyVault = process.env['OBJECT_ID_KEYVAULT_OPERATIONS'];
var keyVaultSp = process.env['SP_KEYVAULT_OPERATIONS'];
```

So to fix this, should I set only `OBJECT_ID_KEYVAULT_OPERATIONS` and not `SP_KEYVAULT_OPERATIONS`?

**Comment by balajikris** [comment]
@shpros , I would try @pomortaz 's suggestion to see if that works.

Although, I would accomplish by editing the code [here](https://github.com/Azure-Samples/key-vault-node-getting-started/blob/master/index.js#L228)
Leave the environment variables as is, because I think you still need the SP to handshake with keyvault [here](https://github.com/Azure-Samples/key-vault-node-getting-started/blob/master/weather-app.js#L78) 

I'm still trying to understand what might be going on here and I'm only expanding on @pomortaz 's suggestion, as to how to go through with that.

One suggestion I'd like to throw out rightaway is see if you could browse the keyvault resource on azure portal and see what service principals have access to the vault and what the access policies are. 
Assuming that assigning access policy is indeed our problem, lets say you were able to manually accomplish this from the azure portal, then running the weather-app sample should just work. Once we figure out what the missing step is, we could incorporate it back in the sample as a trivial last step.

For what its worth, the app did work for me, multiple times when I wrote and tested the sample. So, I'm curious to find out why this won't work or which step wasn't quite clear.

**Comment by shpros** [comment]
@balajikris Commenting out line 229 `applicationId: keyVaultSp,` worked for me. I am able to call weather-app successfully. Thanks so much!

```
current conditions for London,GB: 10.79 Celsius and broken clouds
```

Regarding the rest of your comment, I wasn't able to browse the key vault resource on the Azure portal until I gave appropriate access policy to my real user (my User, not any Application). In any case, the SPs that have access to the vault are listed above under "Details"; search for "get-azurermkeyvault".

I wonder if there has been a change in one of the underlying libraries since you first wrote the tutorial that is causing it to bomb out when both applicationId and objectId are provided.

**Comment by balajikris** [comment]
@shpros - that's awesome. I'm glad we're able to clear this out. Feel free to send a PR to the repo if you like, otherwise, I'll update this when I get some time 😄 
@pomortaz  - thanks for correctly identifying the problem here. As @shpros points out, I also suspect if something changed recently with KV, as this sample used to work as-is for me.

**Comment by balajikris** [comment]
One other thing that struck me was we didn't warn or error when the user did an incorrect operation, i.e, while assigning a new access policy but rather we fail when the user is trying to access a resource that he thought he had access to. Failing much later in the pipeline makes this hard to debug. My 2 cents.

@pomortaz  - can you take this question as well?

**Comment by pomortaz** [comment]
@balajikris nothing changed in this case from Key Vault side. Setting ApplicationId is not an incorrect operation. It is for compound identity scenario, so you could not warn user without knowing the purpose of the operation.

**Comment by balajikris** [comment]
@pomortaz - thanks for the clarification.
